src/pfe/file-watcher/server/src/controllers/projectsController.ts

```diff
@@ -627,13 +613,13 @@ function checkInProgressBuilds(): void {
 async function emitProjectRanks(): Promise<void> {
     logger.logTrace("Emitting project ranks");
     await lock.acquire("buildQueueLock", async done => {
-        buildQueue = buildQueue.filter(temp => {
-            return temp != undefined;
-        });
-        buildQueue.forEach( async (project, index) => {
+        // convert the build queue set into an inplace array for project rank emition that relies on the element index
+        // sets are not great for storing elements in any order, so this is required for the project ranks which relies on position
+        const buildQueueArray = [...buildQueue];
+        buildQueueArray.forEach( async (project, index) => {
```

Comment:
I thought we discovered that sets in JS stay in insertion order which is why we can use it as a queue with unique values. Perhaps the actual problem is the use of `forEach` here? If we iterate over the array with a standard for loop instead and use `buildQueue.values().next()` in each iteration it should work? Then we don't have the overhead of converting from set to array.

Reply:
The insertion do stay in order however there is no indexing in sets. So we can't just do `buildQueue[0]` to get the first element. Since for ranking we need to know the exact position (number) of them, we need to transform it into an array.

The way you described `forEach` don't return the index on sets. It will only give me the element. And sets don't have an easy way to get the "index" of an element

Reply:
This is what I was suggesting:
```
const set1 = new Set();
set1.add('projectA');
set1.add('projectB');

const iterator1 = set1.values();

for(let i=1; i<(set1.size+1); i++) {
  console.log(iterator1.next().value + ' ' + i + '/' + set1.size);  
}
```

Output:
```
> "projectA 1/2"
> "projectB 2/2"
```
It's more efficient than copying to an array and then iterating.

Reply:
Applied those changes and tested it out. PTAL
![Screen Shot 2020-02-13 at 12 26 32 PM](https://user-images.githubusercontent.com/15173354/74461600-c5295700-4e5c-11ea-9868-2aed46f302c5.png)